File: app/src/main/java/com/baikas/sporthub6/firebase/dao/TerrainAddressFirebaseDao.java
package com.baikas.sporthub6.firebase.dao;

import com.baikas.sporthub6.models.TerrainAddress;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.functions.FirebaseFunctions;
import com.google.firebase.functions.HttpsCallableResult;
import com.google.gson.Gson;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

public class TerrainAddressFirebaseDao {
    private final FirebaseFirestore firestore;
    private final String mainCollection = "userSavedTerrainAddresses";

    private final FirebaseFunctions functions;


    @Inject
    public TerrainAddressFirebaseDao(FirebaseFirestore firestore,FirebaseFunctions functions) {
        this.firestore = firestore;
        this.functions = functions;

    }

    public Task<DocumentSnapshot> getTerrainAddresses(String userId){
        DocumentReference docRef = firestore.collection(mainCollection).document(userId);

        return docRef.get();
    }

    public Task<HttpsCallableResult> updateTerrainAddresses(Map<String, List<TerrainAddress>> terrainAddressesMap, String yourId){

        Map<String,Object> map = new HashMap<>();

        map.put("yourId",yourId);

        Gson gson = new Gson();
        String jsonStringTerrainAddressesMap = gson.toJson(terrainAddressesMap);
        map.put("terrainAddressesMap",jsonStringTerrainAddressesMap);


        return functions.getHttpsCallable("terrainRepositoryUpdateTerrain")
                .call(map);
    }

}

File: functions/src/firebasedao/TerrainAddressFirebaseDao.ts
/* eslint-disable */


import {TerrainAddress} from "../models/TerrainAddress";
import * as admin from "firebase-admin";

export class TerrainAddressFirebaseDao {


    public saveTerrainAddress(terrainAddressMap: Map<string, TerrainAddress[]>, yourId: string): Promise<FirebaseFirestore.WriteResult> {

        const doc = admin.firestore().collection("userSavedTerrainAddresses").doc(yourId);

        const record: Record<string, {}> = {};
        terrainAddressMap.forEach((value, key) => {

            const toPlain: {}[] = [];
            value.forEach((terrain: TerrainAddress) => {
                toPlain.push(terrain.toObject());
            });


            record[key] = toPlain;
        });

        return doc.set(record);
    }

    public deleteAllTerrainsForThatUser(userId: string): Promise<FirebaseFirestore.WriteResult> {

        const doc = admin.firestore().collection("userSavedTerrainAddresses").doc(userId);

        return doc.delete();
    }

}
